Add New Unmanic Integration

JeffResc opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
This proposed change adds an integration for [Unmanic](https://docs.unmanic.app/) (see [GitHub](https://github.com/Unmanic/unmanic)) to view various stats (as sensor entities) from your Unmanic server.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: [homeassistant/home-assistant.io#21405](https://github.com/home-assistant/home-assistant.io/pull/21405)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

iMicknl reviewed (changes requested):
For your initial PR, please target a single platform. (e.g. sensor). You can add the other platforms in follow up PR's, where you add one platform per PR.

This will help speed up the review.

JeffResc commented:
> For your initial PR, please target a single platform. (e.g. sensor). You can add the other platforms in follow up PR's, where you add one platform per PR.

Thank you for the suggestion, I've updated this PR to only include the sensor entity. I've updated the documentation PR to reflect this as well.

JeffResc commented:
Just as a reference and to solve any confusion, [the last commit](https://github.com/home-assistant/core/pull/65279/commits/2abe9cc7ab143644cc98d1af5e6c9b61cd00eaed) should have said entities, not integrations.

JeffResc commented on homeassistant/components/unmanic/translations/en.json:
```diff
@@ -0,0 +1,21 @@
+            "cannot_connect": "Unable to connect to Unmanic",
+            "unknown": "Unknown error occured when configuring Unmanic",
+            "already_configured": "An instance of Unmanic has already been configured on this host and port"
+        }
```
I didn't realize this was an option, thank you. I've fixed this now 👍

JeffResc commented:
I'm currently working on the pylint errors, I'll update here when those are all passing

iMicknl commented on homeassistant/components/unmanic/__init__.py:
```diff
@@ -0,0 +1,115 @@
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    for platform in PLATFORMS:
+        if entry.options.get(platform, True):
```
Why is this an option?

JeffResc replied:
Honestly, I'm not too sure why this was written as an option, but I used it from [integration_blueprint](https://github.com/custom-components/integration_blueprint/blob/8b3b9d380aac2b3473d5811b9450b9a412d5fe42/custom_components/integration_blueprint/__init__.py#L57-L62) and that's why it's there still. I tried changing this myself, but I couldn't get a working solution.

iMicknl replied:
What if you just remove it?

JeffResc replied:
If I remove it, all entities show as unavailable

JeffResc replied:
Wait, I forgot to import it and somehow didn't get an import error, this is all set

iMicknl commented on homeassistant/components/unmanic/__init__.py:
```diff
@@ -0,0 +1,115 @@
+    for platform in PLATFORMS:
+        if entry.options.get(platform, True):
+            coordinator.platforms.append(platform)
+            hass.async_add_job(
+                hass.config_entries.async_forward_entry_setup(entry, platform)
+            )
```
```suggestion
            hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

Why not use this one? I am not sure what yours does, but I thought it was legacy.

iMicknl commented on homeassistant/components/unmanic/__init__.py:
```diff
@@ -0,0 +1,115 @@
+    return unloaded
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
```
Why is this required?

JeffResc replied:
Not sure, I thought that made it display the new value when an update was made, but it seems that's not the case. Removed 👍

iMicknl commented on homeassistant/components/unmanic/__init__.py:
```diff
@@ -0,0 +1,115 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle removal of an entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    unloaded = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+                if platform in coordinator.platforms
+            ]
+        )
+    )
+    if unloaded:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unloaded
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
```

You would need to test, but I think you can simplify this a lot with the new helpers.

iMicknl commented on homeassistant/components/unmanic/__init__.py:
```diff
@@ -0,0 +1,115 @@
+            data["version"] = await self.api.get_version()
+            data["workers_status"] = await self.api.get_workers_status()
+            return data
+        except Exception as exception:
```
Do you have any other (more specific) exceptions that you can catch? I don't think you need to catch a broad exception in your `_async_update_data`, since this the Data Update Coordinator takes care of this already.

iMicknl commented on homeassistant/components/unmanic/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+        """Return true if credentials is valid."""
+
+        try:
+            session = async_create_clientsession(self.hass)
```
Why do you create a new session where, while you reuse the Home Assistant session in `__init__`?

JeffResc replied:
How can the one in `__init__` be used? Doesn't `_test_connection()` get called before setting up the entry, then once the entry is setup, then `async_setup_entry()` is called where `session` is created?

iMicknl replied:
You use two different functions to get a (new) session. Create vs get session. You could use get in both implementations.

JeffResc replied:
Do you mean literally switch `session = async_create_clientsession(self.hass)` for `session = async_get_clientsession(self.hass)`? If so, that creates an error with "Failed to connect" which I believe is a generic exception from aiohttp, likely due to a missing session.

JeffResc replied:
I forgot to import `async_get_clientsession` and it somehow didn't raise an import error. Anyways, it's all set now 👍

iMicknl commented on homeassistant/components/unmanic/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+        self._errors = {}
+
+        if user_input is not None:
+            configuration_already_exists = await self._config_exists(
+                user_input[CONF_HOST],
+                user_input[CONF_PORT],
+                self.hass,
+            )
```
Why manually check if the configuration already exists? You can set the unique id and use the helpers. My only concern would be that a host is not a super unique address, is there nothing in the config / setup that is more unique?

```suggestion
            await self.async_set_unique_id(user_input[CONF_HOST] + user_input[CONF_PORT])

            self._abort_if_unique_id_configured()
```

JeffResc replied:
With this change, when it's aborted due to a duplication configuration, it looks like this:
![Aborted](https://i.imgur.com/GkclOM0.png)

Is there any way to catch this as an exception or something similar and display a more normal-looking error, similar to how it was before (or how the `cannot_connect` error looks)?
![Cannot connect error](https://i.imgur.com/XcAPc2a.png)

iMicknl commented on homeassistant/components/unmanic/const.py:
```diff
@@ -0,0 +1,16 @@
+DEFAULT_NAME = "Unmanic"
+
+# Platforms
+SENSOR = "sensor"
+PLATFORMS = [SENSOR]
```
```suggestion
PLATFORMS = [Platform.SENSOR]
```

and ```from homeassistant.const import Platform```

iMicknl commented on homeassistant/components/unmanic/const.py:
```diff
@@ -0,0 +1,16 @@
+# Base component constants
+NAME = "Unmanic"
+DOMAIN = "unmanic"
+DOMAIN_DATA = f"{DOMAIN}_data"
```
Where do you use this?

iMicknl commented on homeassistant/components/unmanic/entity.py:
```diff
@@ -0,0 +1,33 @@
+        configuration_url += f"{self.coordinator.api.host}:{self.coordinator.api.port}"
+        configuration_url += "/unmanic/ui/dashboard/"
+
+        return {
+            "identifiers": {(DOMAIN, self._device_id)},
+            "name": f"{NAME} ({self.coordinator.api.host})",
+            "model": NAME,
+            "sw_version": self.coordinator.data["version"],
+            "manufacturer": NAME,
+            "configuration_url": configuration_url,
+        }
```
Please use `DeviceInfo` from `from homeassistant.helpers.entity import DeviceInfo`

iMicknl commented on homeassistant/components/unmanic/entity.py:
```diff
@@ -0,0 +1,33 @@
+    @property
+    def device_info(self):
+        """Unmanic device information."""
+        if self._device_id is None:
```
When can this happen? Does this mean that the entity is not set up?

JeffResc replied:
Yes, this used to mean it in an older iteration of the integration I was working on. It no longer seems necessary and I've removed it

iMicknl commented on homeassistant/components/unmanic/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/unmanic",
+  "requirements": ["unmanic-api==0.0.5"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
```suggestion
```

iMicknl commented on homeassistant/components/unmanic/sensor.py:
```diff
@@ -0,0 +1,130 @@
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        if self.key == "total-workers":
```
I am not a huge fan of all the `if/else` statements in this entity. Are you familiar with the Entity Descriptions? See tuya for an example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/tuya/sensor.py

iMicknl commented on homeassistant/components/unmanic/sensor.py:
```diff
@@ -0,0 +1,130 @@
+    @property
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        attrs = {}
```
Why do you need all these extra state attributes? Shouldn't they just be extra entities?

JeffResc replied:
At any moment, there could be theoretically unlimited in the history or queue. This allows a quick look at the last 10 items in either the history or the queue. While they could be their own entities and I don't believe this is a great use of extra state attributes, I opted not to do that to not clutter up with 20 additional entities. What do you think is the best way I could represent this?

iMicknl reviewed (changes requested):
Had a quick look to get you started. Would be good to mark it as a draft until CI/CD is fully green :-).

JeffResc commented:
While attempting to squash the commits in the branch for this PR, I accidentally destroyed the branch past easy repair. For that reason, I have moved this PR to #65352.
